napari/components/_tests/test_layer_slicer.py

```diff
@@ -271,6 +275,90 @@ def test_slice_layers_async_task_to_layers_lock(layer_slicer):
     assert task not in layer_slicer._layers_to_task
 
 
+def test_slice_layers_exception_main_thread(layer_slicer):
+    """Exception is raised on the main thread from an error on the main
+    thread immediately when the task is created."""
+
+    class FakeAsyncLayerError(FakeAsyncLayer):
+        def _make_slice_request(self, dims) -> FakeSliceRequest:
+            raise RuntimeError('_make_slice_request')
+
+    layer = FakeAsyncLayerError()
+    with pytest.raises(RuntimeError, match='_make_slice_request'):
+        layer_slicer.slice_layers_async(layers=[layer], dims=Dims())
+
+
+def test_slice_layers_exception_subthread_on_result(layer_slicer):
+    """Exception is raised on the main thread from an error on a subthread
+    only after result is called, not upon submission of the task."""
+
+    class FakeAsyncLayerError(FakeAsyncLayer):
+        def _get_slice(self, request: FakeSliceRequest) -> FakeSliceResponse:
+            raise RuntimeError('_get_slice')
+
+    layer = FakeAsyncLayerError()
+    future = layer_slicer.slice_layers_async(layers=[layer], dims=Dims())
+
+    done, _ = wait([future], timeout=5)
+    if done:
+        with pytest.raises(RuntimeError, match='_get_slice'):
+            future.result()
+    else:
+        raise TimeoutError('Test future did not complete within timeout.')
+
+
+def test_wait_until_idle(layer_slicer, single_threaded_executor):
+    dims = Dims()
+    layer = FakeAsyncLayer()
+
+    with layer.lock:
+        slice_future = layer_slicer.slice_layers_async(
+            layers=[layer], dims=dims
+        )
+        _wait_until_running(slice_future)
+        # The slice task has started, but has not finished yet
+        # because we are holding the layer's slicing lock.
+        assert len(layer_slicer._layers_to_task) > 0
+        # We can't call wait_until_idle on this thread because we're
+        # holding the layer's slice lock, so submit it to be executed
+        # on another thread and also wait for it to start.
+        wait_future = single_threaded_executor.submit(
+            layer_slicer.wait_until_idle
+        )
+        _wait_until_running(wait_future)
+
+    wait_future.result()
+    assert len(layer_slicer._layers_to_task) == 0
+
+
+def _wait_until_running(future: Future):
+    while not future.running():
+        continue
```

Comment:
I know I wrote this, but we should probably throttle this looping check a little.

```suggestion
        time.sleep(0.01)
```

Alternatively, maybe we'll get better ideas from others in other futures.

Reply:
done.